ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java

```diff
@@ -1118,46 +1037,34 @@ private RequestVoteReplyProto requestVote(
     synchronized (this) {
       // Check life cycle state again to avoid the PAUSING/PAUSED state.
       assertLifeCycleState(LifeCycle.States.RUNNING);
-      FollowerState fs = role.getFollowerState().orElse(null);
-      if (shouldWithholdVotes(candidateTerm)) {
-        LOG.info("{}-{}: Withhold vote from candidate {} with term {}. State: leader={}, term={}, lastRpcElapsed={}",
-            getMemberId(), role, candidateId, candidateTerm, state.getLeaderId(), state.getCurrentTerm(),
-            fs != null? fs.getLastRpcTime().elapsedTimeMs() + "ms": null);
-      } else if (state.recognizeCandidate(candidateId, candidateTerm)) {
-        final boolean termUpdated = changeToFollower(candidateTerm, true, "recognizeCandidate:" + candidateId);
-        // if current server is leader, before changeToFollower FollowerState should be null,
-        // so after leader changeToFollower we should get FollowerState again.
-        fs = role.getFollowerState().orElse(null);
-
-        // see Section 5.4.1 Election restriction
-        RaftPeer candidate = getRaftConf().getPeer(candidateId);
-        if (fs != null && candidate != null) {
-          int compare = ServerState.compareLog(state.getLastEntry(), candidateLastEntry);
-          int priority = getRaftConf().getPeer(getId()).getPriority();
-          LOG.info("{} priority:{} candidate:{} candidatePriority:{} compare:{}",
-              this, priority, candidate, candidate.getPriority(), compare);
-          // vote for candidate if:
-          // 1. log lags behind candidate
-          // 2. log equals candidate's, and priority less or equal candidate's
-          if (compare < 0 || (compare == 0 && priority <= candidate.getPriority())) {
-            state.grantVote(candidateId);
-            voteGranted = true;
+
+      final VoteContext context = new VoteContext(this, phase, candidateId);
+      final RaftPeer candidate = context.recognizeCandidate(candidateTerm);
+      LOG.info("{}: recognizeCandidate {}? {}", getMemberId(), candidateId, candidate != null);
+      if (candidate != null) {
```

Comment:
Can we put  `voteGranted = context.shouldVote(candidate, candidateLastEntry);` here ? Because Both vote and pre-vote need this check.

Reply:
That's a good idea.  Will update the code.  Thanks.